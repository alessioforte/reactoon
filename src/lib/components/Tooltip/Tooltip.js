import React from 'react'
import styled, { keyframes } from 'styled-components'

export default ({ children, text }) => {
    return (
        <Popup>
            <Info>{ text }</Info>
            { children }
        </Popup>
    )
}

const delay = keyframes`
    0% {
        opacity: 0;
    }
    75% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
`
const Popup = styled.div`
    position: relative;
    display: inline-block;
    & > div:first-child {
            display: none;
            animation: 1.5s ${delay} ease;
        }
    &:hover {
        & >div:first-child {
            display: block;

        }
    }
`
const Info = styled.div`
    box-sizing: border-box;
    position: absolute;
    background: #1a1a1a;
    padding: 9px;
    bottom: 45px;
    border-radius: 3px;
    font-size: 12px;
    left: calc(50% - 60px);
    width: 120px;
    text-align: center;
    box-shadow: 0px 1px 30px 1px rgba(0, 0, 0, 0.2);
    &::after {
        content: '';
        width: 15px;
        height: 15px;
        background: #1a1a1a;
        position: absolute;
        bottom: -7px;
        left: 53px;
        transform: rotate(45deg);
        border-radius: 15px 1px 2px 1px;
    }
`