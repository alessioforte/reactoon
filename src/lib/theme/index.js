/**
 * Default Theme
 */

const colors = {
    primary: '#ffff00',
    secondary: '#DAA520',
    action: '',
    success: '#00FA9A',
    warning: '#FFA500',
    danger: '#B22222',
    error: '#B22222',
    info: '#00BFFF',
    light: '#fff',
    dark: '#1a1a1a',
    text: '#fff',
    textLight: '#ffffff',
    textDark: '#000000',
    textIdle: '',
    textActive: '',
    textError: '',
    groundzero: '#EFEFEF',
    background: '#eee',
    flatground: '#D3D3D3',
    ground: '#B5B5B5',
    hillground: '#848484',
    highground: '#3A3A3A',
    active: '#FFD700',
    selected: '#999',
    hover: '#2182BD',
    idle: '#888',
    disabled: '#778899',
    focus: 'none',
    holiday: '#FF4365',
    shadow: '0px 1px 5px 1px rgba(0, 0, 0, 0.2)'
}

const borders = {
    radius: 0
}

export const styles = {
    colors,
    borders
}

export function getContrastYIQ(hexcolor) {
    if (typeof hexcolor !== 'string') throw new Error(`${hexcolor} is not a string`)
    if (hexcolor[0] !== '#') throw new Error(`${hexcolor} - HEX color not valid`)

    hexcolor = hexcolor.replace('#', '');

    if (hexcolor.length < 3) throw new Error(`${hexcolor} - HEX color not valid`)
    if (hexcolor.length === 3) hexcolor = hexcolor + hexcolor

    var r = parseInt(hexcolor.substr(0,2),16);
    var g = parseInt(hexcolor.substr(2,2),16);
    var b = parseInt(hexcolor.substr(4,2),16);

    var yiq = ((r*299)+(g*587)+(b*114))/1000;

    return (yiq >= 128) ? 'dark' : 'light';
}

export default {
    styles,
    getContrastYIQ
}
