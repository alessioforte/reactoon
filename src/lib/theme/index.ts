/**
 * Default Theme
 */

const COLORS = {
  primary: '#ffff00',
  secondary: '#DAA520',
  success: '#00FA9A',
  warning: '#FFA500',
  danger: '#B22222',
  error: '#B22222',
  info: '#00BFFF',
  light: '#fff',
  dark: '#1a1a1a',
  // text: '#fff',
  // textLight: '#ffffff',
  // textDark: '#000000',
  // textIdle: '',
  // textActive: '',
  // textError: '',

  groundzero: '#EFEFEF',
  background: '#eee',
  flatground: '#D3D3D3',
  ground: '#B5B5B5',
  upperground: '#848484',
  higherground: '#3A3A3A',

  active: '#FFD700',
  selected: '#999',
  hover: '#2182BD',
  idle: '#888',
  disabled: '#778899',
  focus: '#61dafb',
  holiday: '#FF4365',
  shadow: '0px 1px 5px 1px rgba(0, 0, 0, 0.2)'
};

const BORDER = {
  radius: 0,
  width: '1px'
};

const SIZE = {
  is: 'xs',
  xs: {
    height: '',
    font: ''
  },
  sm: {},
  md: {},
  lg: {},
  xl: {}
};

export const styles = {
  colors: COLORS,
  border: BORDER,
  size: SIZE
};

export function create(theme) {
  let colors = COLORS;
  let border = BORDER;
  if (theme.colors) colors = { ...colors, ...theme.colors };
  if (theme.border) border = { ...border, ...theme.border };
  return {
    colors,
    border
  };
}

export function getContrastYIQ(hex): string {
  const color = validateHex(hex);
  if (!color) {
    console.warn(`Something goes wrong getContrastYIQ not applied`);
    return 'dark';
  }
  const [r, g, b] = colorHexToRgb(color);
  const yiq = (r * 299 + g * 587 + b * 114) / 1000;
  return yiq >= 128 ? 'dark' : 'light';
}

export function getColorLuminance(hex, luminance): string {
  const color = validateHex(hex);
  if (!color) {
    console.warn(`Something goes wrong getColorLuminance not applied`);
    return hex;
  }
  const rgb = colorHexToRgb(color);
  hex = '#';
  luminance = luminance || 0;
  rgb.forEach(c => {
    const x = Math.round(Math.min(Math.max(0, c + c * luminance), 255)).toString(
      16
    );
    hex += ('00' + x).substr(x.length);
  });
  return hex;
}

function validateHex(color) {
  if (typeof color !== 'string') {
    console.error(`Validate hex color: ${color} is not a string`);
    return false;
  }
  if (color[0] !== '#') {
    console.error(`Validate hex color: ${color} is not a valid hex color`);
    return false;
  }
  color = color.replace('#', '');
  if (color.length < 3) {
    console.error(`Validate hex color: #${color} is not a valid hex color`);
    return false;
  }
  if (color.length === 3) {
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }
  return `#${color}`;
}

function colorHexToRgb(hex) {
  const r = parseInt(hex[1] + hex[2], 16);
  const g = parseInt(hex[3] + hex[4], 16);
  const b = parseInt(hex[5] + hex[6], 16);
  return [r, g, b];
}

export default {
  create,
  styles,
  getContrastYIQ,
  getColorLuminance
};
