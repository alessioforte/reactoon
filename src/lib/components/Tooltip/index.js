import React, { Children, useState, useEffect } from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';
import { styles, getContrastYIQ } from '../../theme';

const ROOT_NODE = document.createElement('div');

const Tooltip = ({ render, children }) => {
  const target = React.createRef();
  const tip = React.createRef();

  const [visible, setState] = useState(false);

  useEffect(() => {
    if (tip.current) {
      const rect = target.current.getBoundingClientRect();
      const { innerHeight, innerWidth, scrollY } = window;
      const { width, height } = tip.current.getBoundingClientRect();
      const right = innerWidth - (rect.x + rect.width);
      const position = {
        bottom: `${innerHeight - rect.top - scrollY + 5}px`,
        left: `${rect.left + rect.width / 2 - width / 2}px`
      };

      if (rect.x < width / 2 && width > rect.width) {
        position.left = `${rect.left}px`;
      }

      if (right < width / 2) {
        position.left = `${innerWidth - (innerWidth - rect.x) - width + rect.width}px`;
      }

      if (rect.y < height) {
        position.bottom = `${innerHeight - rect.top - scrollY - height - rect.height - 5}px`;
      }

      Object.assign(tip.current.style, position);
    }
  });

  const show = () => {
    setState(true);
  };

  const hide = () => {
    setState(false);
  };

  const renderTooltip = () => {
    return createPortal(<Tip ref={tip}>{render}</Tip>, ROOT_NODE);
  };

  return (
    <>
      <Target
        onMouseEnter={show}
        onMouseLeave={hide}
        onFocus={show}
        onBlur={hide}
        ref={target}
        aria-haspopup='true'
      >
        {Children.toArray(children)}
      </Target>
      {visible && renderTooltip()}
    </>
  );
};

function setRoot(APP_NODE, id) {
  ROOT_NODE.setAttribute('id', id);
  APP_NODE.insertAdjacentElement('afterend', ROOT_NODE);
}

Tooltip.propTypes = {
  children: PropTypes.node.isRequired,
  render: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  theme: PropTypes.object
};

Tooltip.defaultProps = {
  theme: styles
};

Tooltip.setRoot = setRoot;

export default Tooltip;

export const delay = keyframes`
    0% {
        opacity: 0;
    }
    75% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
`;
export const Target = styled.div`
  display: inline-block;
`;
export const Tip = styled.div`
  position: absolute;
  left: -800px;
  padding: 9px;
  animation: 1.2s ${delay} ease;
  font-size: 12px;
  max-height: 150px;
  max-width: 300px;
  box-sizing: border-box;
  text-align: center;
  background: ${props => props.theme.colors.background};
  color: ${props =>
    props.theme.colors[getContrastYIQ(props.theme.colors.background)]};
  box-shadow: ${props => props.theme.colors.shadow};
  border-radius: ${props => props.theme.border.radius + 'px'};
`;
