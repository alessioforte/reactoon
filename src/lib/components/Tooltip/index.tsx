import React, {
  Children,
  useState,
  useRef,
  useEffect,
  ReactNode,
  FunctionComponent,
  ForwardRefExoticComponent
} from 'react';
import { createPortal } from 'react-dom';
import styled, { keyframes, withTheme } from 'styled-components';
import { styles, getContrastYIQ } from '../../theme';

let ROOT_ID = 'root-tooltip';

type Props = {
  children: React.ReactElement;
  content: (props: any) => ReactNode;
  theme: any;
};

const Tooltip: FunctionComponent<Props> = ({ content, children, theme = styles }) => {
  const target: React.MutableRefObject<HTMLDivElement | undefined> = useRef();
  const tip: React.MutableRefObject<HTMLDivElement | undefined> = useRef();

  const [visible, setState] = useState<boolean>(false);

  useEffect(() => {
    if (tip.current && target.current) {
      const rect = target.current.getBoundingClientRect();
      const { innerHeight, innerWidth, scrollY } = window;
      const { width, height } = tip.current.getBoundingClientRect();
      const right = innerWidth - (rect.x + rect.width);
      const position = {
        bottom: `${innerHeight - rect.top - scrollY + 5}px`,
        left: `${rect.left + rect.width / 2 - width / 2}px`
      };

      if (width / 2 > rect.x + rect.width / 2 && width > rect.width) {
        position.left = `${rect.left}px`;
      }

      if (right < width / 2) {
        position.left = `${innerWidth -
          (innerWidth - rect.x) -
          width +
          rect.width}px`;
      }

      if (rect.y < height) {
        position.bottom = `${innerHeight -
          rect.top -
          scrollY -
          height -
          rect.height -
          5}px`;
      }

      Object.assign(tip.current.style, position);
    }
  });

  const show = () => {
    setState(true);
  };

  const hide = () => {
    setState(false);
  };

  const renderTooltip = () => {
    const ROOT_NODE: any = document.getElementById(ROOT_ID);
    const Root = <Tip ref={tip}>{content}</Tip>;
    return createPortal(Root, ROOT_NODE);
  };

  return (
    <>
      <Target
        onMouseEnter={show}
        onMouseLeave={hide}
        onFocus={show}
        onBlur={hide}
        ref={target}
        aria-haspopup='true'
      >
        {Children.toArray(children)}
      </Target>
      {visible && renderTooltip()}
    </>
  );
};

interface FC<P> extends ForwardRefExoticComponent<P> {
  setRoot: (APP_NODE: Element, id: string) => void;
}

const FREC: FC<Props> = withTheme(Tooltip) as FC<Props>;

FREC.setRoot = (APP_NODE, id) => {
  ROOT_ID = id;
  let node = document.getElementById(ROOT_ID);
  if (!node) {
    node = document.createElement('div');
    node.setAttribute('id', ROOT_ID);
    APP_NODE.insertAdjacentElement('afterend', node);
  }
};

export default FREC;

export const delay = keyframes`
    0% {
        opacity: 0;
    }
    75% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
`;
export const Target = styled.div<{ ref: any }>`
  display: inline-block;
  width: fit-content;
`;
export const Tip = styled.div<{ ref: any }>`
  position: absolute;
  left: -800px;
  z-index: 99;
  padding: 9px;
  animation: 1.2s ${delay} ease;
  font-size: 12px;
  max-height: 150px;
  max-width: 300px;
  box-sizing: border-box;
  text-align: center;
  background: ${props => props.theme.colors.background};
  color: ${props =>
    props.theme.colors[getContrastYIQ(props.theme.colors.background)]};
  box-shadow: ${props => props.theme.colors.shadow};
  border-radius: ${props => props.theme.border.radius + 'px'};
`;
