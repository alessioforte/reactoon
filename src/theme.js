import { Theme } from './lib';

const theme = Theme.create({
  colors: {
    primary: '#1FA588',
    secondary: '#86C5E0',
    action: '#F7AC91',
    active: '#FFE1C4',
    selected: '#008080',
    hover: '#1FA588',
    success: '#00FA9A',
    warning: '#FFA500',
    danger: '#B22222',
    error: '#B22222',
    info: '#00BFFF',
    light: '#fff',
    dark: '#1a1a1a',
    text: 'red',
    textLight: '#ffffff',
    textDark: '#000000',
    textIdle: '',
    textActive: '',
    textError: '',
    groundzero: '#FFFFFF',
    background: '#F9F9F9',
    flatground: '#EAEAEA',
    ground: '#CCCCCC',
    hillground: '#707070',
    highground: '#303030',
    idle: '#eee',
    disabled: '#778899',
    focus: '#fafafa',
    holiday: '#FF4365',
    shadow: '0px 1px 5px 1px rgba(0, 0, 0, 0.2)'
  },
  borders: {
    radius: 5
  }
});

export default theme;
