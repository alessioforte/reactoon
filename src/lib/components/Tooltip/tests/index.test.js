import React from 'react'
import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Tooltip, { Target, Tip } from '../index'
import { styles } from '../../../theme'

describe('<Tooltip />', () => {
  it('should render and match the snapshot', () => {
    const { container: { firstChild } } = render(<Tooltip content='message'>TEST</Tooltip>)
    expect(firstChild).toMatchSnapshot()
  })
})

describe('<Target /> styled component', () => {
    it('should have display: inline-block;', () => {
      const tree = renderer.create(<ThemeProvider theme={styles}><Target /></ThemeProvider>).toJSON()
      expect(tree).toHaveStyleRule('display', 'inline-block')
    })
  })

describe('<Tip /> styled component', () => {
  it('should have max-width: 300px; max-height: 150px', () => {
    const tree = renderer.create(<ThemeProvider theme={styles}><Tip /></ThemeProvider>).toJSON()
    expect(tree).toHaveStyleRule('max-width', '300px')
    expect(tree).toHaveStyleRule('max-height', '150px')
  })
})