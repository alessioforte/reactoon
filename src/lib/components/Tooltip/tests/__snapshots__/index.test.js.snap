// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> should render and match the snapshot 1`] = `
.c0 {
  display: inline-block;
}

<div
  class="c0"
>
  TEST
</div>
`;
